DirectProgramming/DPC++FPGA/Tutorials/Features/read_only_cache/README.md

```diff
@@ -0,0 +1,261 @@
+Guide](https://software.intel.com/en-us/oneapi-programming-guide) is a general
+resource for target-independent DPC++ programming.
+
+| Optimized for                     | Description
+---                                 |---
+| OS                                | Linux* Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE* 15; Windows* 10
+| Hardware                          | Intel® Programmable Acceleration Card (PAC) with Intel Arria® 10 GX FPGA <br> Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX) <br> Intel® FPGA 3rd party / custom platforms with oneAPI support <br> *__Note__: Intel® FPGA PAC hardware is only compatible with Ubuntu 18.04* 
+| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit 
+| What you will learn               | How and when to use the read-only cache feature
+| Time to complete                  | 30 minutes
+
+
+
+## Purpose
+
+This FPGA tutorial demonstrates an example of using the read-only cache to
+boost the throughput of an FPGA DPC++ design. Specifically, the read-only cache
+is most appropriate for table lookups that are constant throughout the
+execution of a kernel and is optimized for high cache hit performance.
+
+To enable the read-only cache, the `-Xsread-only-cache-size<N>` flag should be
+passed to the `dpcpp` command. Each kernel will get its own *private* version
+of the cache that serves all reads in the kernel from read-only no-alias
+accessors. Read-only no-alias accessors are accessors that have both the
+`read_only` and the `no_alias` properties:
+```c++
+accessor sqrt_lut(sqrt_lut_buf, h, read_only, accessor_property_list{no_alias});
+```
+The `read_only` property is required because the cache is *read_only*. The
+`no_alias` property is required to guarantee that the buffer will not be
+written to from the kernel through another accessor or through a USM pointer.
+Note that the same no-alias behavior can be achieved using the kernel attribute
+`[[intel::kernel_args_restrict]]`, when applicable.
+
+Each private cache is also replicated as many times as needed so that it can
+expose extra read ports. The size of each replicate is `<N>` bytes as specified
+by the `-Xsread-only-cache-size=<N>` flag.
+
+### Tutorial Design 
+The basic function performed by the tutorial kernel is a series of table
+lookups from a buffer (`sqrt_lut_buf`) that contains the square root values of
+the first 512 integers. By default, the compiler will generate load-store units
+(LSUs) that are optimized for the case where global memory accesses are
+contiguous. When the memory accesses are non-contiguous, like it is the case
+with this tutorial design, these LSUs tend to suffer major throughput loss. The
+read-only cache can sometimes help in such situations, especially when sized
+correctly.
+
+This tutorial requires compiling the source code twice: once with the
+`-Xsread-only-cache-size=<N>` flag and once without it. Because the look-up
```

Comment:
In general, tutorials that only need to be compiled once are preferred.  Is it possible to create two kernels, one that uses the read only cache and one that does not, and have both kernels in a single compile?

Reply:
I agree and I wanted to do that. However, all feature tutorials where the feature is enabled using a flag will require two separate compiles which is unfortunate. One way around this, for this particular tutorial, is to have two kernels where one uses a `read_write` accessor from the LUT (=> no cache) and another that uses a `read_only` accessor from the LUT (=> cache) but that's an unfair comparison since `read_write` adds extra overhead v.s. `read_only` that is unrelated to caching.

Btw, I am working on another tutorial for `-Xsno-interleaving` that has the same problem.